File: studentassistant/src/main/java/ru/erdenian/studentassistant/fragment/SchedulePageFragment.java
package ru.erdenian.studentassistant.fragment;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.google.common.base.Joiner;

import org.joda.time.LocalDate;

import java.util.List;

import ru.erdenian.studentassistant.R;
import ru.erdenian.studentassistant.schedule.Lesson;
import ru.erdenian.studentassistant.schedule.ScheduleManager;
import ru.erdenian.studentassistant.schedule.Semester;

/**
 * Todo: описание класса.
 *
 * @author Ilya Solovyev
 * @version 0.0.0
 * @since 0.0.0
 */
public class SchedulePageFragment extends Fragment {

    static final String PAGE_SEMESTER_INDEX = "page_semester_index";
    static final String PAGE_DATE = "page_date";

    static final String TIME_FORMAT = "HH:mm";

    private Semester semester;
    private LocalDate day;

    public static SchedulePageFragment newInstance(int semesterIndex, LocalDate date) {
        SchedulePageFragment schedulePageFragment = new SchedulePageFragment();
        Bundle arguments = new Bundle();
        arguments.putInt(PAGE_SEMESTER_INDEX, semesterIndex);
        arguments.putString(PAGE_DATE, date.toString());
        schedulePageFragment.setArguments(arguments);
        return schedulePageFragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        semester = ScheduleManager.getSemesters().asList().get(getArguments().getInt(PAGE_SEMESTER_INDEX));
        day = new LocalDate(getArguments().getString(PAGE_DATE));
    }

    @Override
    public View onCreateView(LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {

        List<Lesson> lessons = semester.getLessons(day);

        if (lessons.size() == 0) {
            return inflater.inflate(R.layout.fragment_free_day, container, false);
        }

        View view = inflater.inflate(R.layout.scroll_view, container, false);
        LinearLayout llCardsParent = (LinearLayout) view.findViewById(R.id.scroll_view_items_parent);

        for (final Lesson lesson : lessons) {
            View card = inflater.inflate(R.layout.card_schedule, llCardsParent, false);

            TextView tvStartTime = (TextView) card.findViewById(R.id.card_schedule_start_time);
            TextView tvEndTime = (TextView) card.findViewById(R.id.card_schedule_end_time);
            TextView tvClassrooms = (TextView) card.findViewById(R.id.card_schedule_classrooms);
            TextView tvType = (TextView) card.findViewById(R.id.card_schedule_type);
            TextView tvName = (TextView) card.findViewById(R.id.card_schedule_name);

            tvStartTime.setText(lesson.getStartTime().toString(TIME_FORMAT));
            tvEndTime.setText(lesson.getEndTime().toString(TIME_FORMAT));

            if (lesson.getClassrooms().size() > 0) {
                tvClassrooms.setText(Joiner.on(", ").join(lesson.getClassrooms()));
            } else {
                card.findViewById(R.id.card_schedule_classrooms_icon).setVisibility(View.GONE);
            }

            if (lesson.getType() != null)
                tvType.setText(lesson.getType());
            else
                tvType.setHeight(0);

            tvName.setText(lesson.getName());

            LinearLayout llTeachersParent =
                    (LinearLayout) card.findViewById(R.id.card_schedule_teachers_parent);

            for (String name : lesson.getTeachers()) {
                View teacher = inflater.inflate(R.layout.textview_teacher, llTeachersParent, false);
                TextView tvTeacher = (TextView) teacher.findViewById(R.id.textview_teacher);
                tvTeacher.setText(name);
                llTeachersParent.addView(teacher);
            }

            card.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Toast.makeText(getContext(), lesson.getName(), Toast.LENGTH_SHORT).show();
                }
            });

            llCardsParent.addView(card);
        }

        return view;
    }
}

File: studentassistant/src/main/kotlin/ru/erdenian/studentassistant/fragment/SchedulePageFragment.kt
package ru.erdenian.studentassistant.fragment

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.*
import android.widget.AdapterView
import android.widget.ListView
import org.jetbrains.anko.alert
import org.jetbrains.anko.startActivity
import org.joda.time.LocalDate
import ru.erdenian.studentassistant.R
import ru.erdenian.studentassistant.activity.COPY
import ru.erdenian.studentassistant.activity.LESSON_ID
import ru.erdenian.studentassistant.activity.LessonEditorActivity
import ru.erdenian.studentassistant.activity.SEMESTER_ID
import ru.erdenian.studentassistant.adapter.ScheduleListAdapter
import ru.erdenian.studentassistant.localdata.ScheduleManager

class SchedulePageFragment : Fragment() {

  companion object {

    private const val TIME_FORMAT = "HH:mm"

    private const val PAGE_SEMESTER_ID = "page_semester_id"
    private const val PAGE_DATE = "page_date"
    private const val PAGE_WEEKDAY = "page_weekday"
    private const val SHOW_WEEKS_AND_DATES = "show_weeks_and_dates"

    fun newInstance(semesterId: Long, date: LocalDate): SchedulePageFragment = SchedulePageFragment().apply {
      arguments = Bundle().apply {
        putLong(PAGE_SEMESTER_ID, semesterId)
        putString(PAGE_DATE, date.toString())
        putBoolean(SHOW_WEEKS_AND_DATES, false)
      }
    }

    fun newInstance(semesterId: Long, weekday: Int): SchedulePageFragment = SchedulePageFragment().apply {
      arguments = Bundle().apply {
        putLong(PAGE_SEMESTER_ID, semesterId)
        putInt(PAGE_WEEKDAY, weekday)
        putBoolean(SHOW_WEEKS_AND_DATES, true)
      }
    }
  }

  private val semesterId: Long by lazy { arguments.getLong(PAGE_SEMESTER_ID, -1L) }
  private val showWeeksAndDates: Boolean by lazy { arguments.getBoolean(SHOW_WEEKS_AND_DATES) }
  private val day: LocalDate by lazy { LocalDate(arguments.getString(PAGE_DATE)) }
  private val weekday: Int by lazy { arguments.getInt(PAGE_WEEKDAY, -1) }

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
    val lessons = if (showWeeksAndDates) ScheduleManager.getLessons(semesterId, weekday)
    else ScheduleManager.getLessons(semesterId, day)

    return if (lessons.isEmpty()) inflater.inflate(R.layout.fragment_free_day, container, false)
    else (inflater.inflate(R.layout.list_view, container, false) as ListView).apply {
      adapter = ScheduleListAdapter(context, semesterId, showWeeksAndDates, lessons)
      if (showWeeksAndDates) registerForContextMenu(this)
    }
  }

  override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenu.ContextMenuInfo?) {
    super.onCreateContextMenu(menu, v, menuInfo)
    activity.menuInflater.inflate(R.menu.context_menu_schedule_page_fragment, menu)
    menu.setHeaderTitle(ScheduleManager.getLesson(semesterId, (menuInfo as AdapterView.AdapterContextMenuInfo).id).subjectName)
  }

  override fun onContextItemSelected(item: MenuItem): Boolean {
    val info = item.menuInfo as AdapterView.AdapterContextMenuInfo
    val lesson = ScheduleManager.getLesson(semesterId, info.id)
    when (item.itemId) {
      R.id.context_menu_schedule_page_fragment_copy -> context.startActivity<LessonEditorActivity>(
          context.SEMESTER_ID to semesterId,
          context.LESSON_ID to lesson.id,
          context.COPY to true
      )
      R.id.context_menu_schedule_page_fragment_delete -> {
        fun remove() {
          ScheduleManager.removeLesson(semesterId, lesson.id)
          //context.startService<ScheduleService>()
        }

        if (ScheduleManager.getHomeworks(semesterId, lesson.subjectName).isNotEmpty() &&
            (ScheduleManager.getLessons(semesterId, lesson.subjectName).size == 1)) {
          context.alert(R.string.activity_lesson_editor_alert_delete_homeworks_message,
              R.string.activity_lesson_editor_alert_delete_homeworks_title) {
            positiveButton(R.string.activity_lesson_editor_alert_delete_homeworks_yes) { remove() }
            negativeButton(R.string.activity_lesson_editor_alert_delete_homeworks_cancel) {}
          }.show()
        } else {
          context.alert(R.string.activity_lesson_editor_alert_delete_message) {
            positiveButton(R.string.activity_lesson_editor_alert_delete_yes) { remove() }
            negativeButton(R.string.activity_lesson_editor_alert_delete_no) {}
          }.show()
        }
      }
      else -> throw IllegalArgumentException("Неизвестный id: ${item.itemId}")
    }
    return true
  }
}
